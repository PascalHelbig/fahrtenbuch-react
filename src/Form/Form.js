import React, { Component, PropTypes } from 'react';
import Formsy from 'formsy-react';

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {
      values: {},
      isValid: false,
    };

    this.handleChanges = this.handleChanges.bind(this);
  }

  handleChanges(values) {
    this.setState({ values });
  }

  render() {
    const { values, isValid } = this.state;
    return (
      <Formsy.Form
        onChange={this.handleChanges}
        onValid={() => this.setState({ isValid: true })}
        onInvalid={() => this.setState({ isValid: false })}
      >
        {this.props.children(values, isValid)}
      </Formsy.Form>
    );
  }
}

Form.propTypes = {
  children: PropTypes.func.isRequired,
};

export default Form;
