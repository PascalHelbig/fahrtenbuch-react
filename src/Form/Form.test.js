import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Formsy from 'formsy-react';
import Form from './Form';

it('renders correctly', () => {
  const form = shallow(<Form>{() => {}}</Form>);
  expect(toJSON(form)).toMatchSnapshot();
});

it('calls the childrens function with its state', () => {
  const children = jest.fn();
  const form = shallow(<Form>{children}</Form>);

  form.setState({
    values: {
      email: 'test@test.de', password: '1234',
    },
    isValid: true,
  });

  expect(children).toHaveBeenLastCalledWith(
    { email: 'test@test.de', password: '1234' },
    true,
  );
});

it('sets state on handleChanges', () => {
  const form = shallow(<Form>{() => {}}</Form>);
  form.instance().handleChanges({ email: 'test@test.de', password: '1234' });

  expect(form.state('values')).toEqual({ email: 'test@test.de', password: '1234' });
});

it('set isValid to true when formsy is valid', () => {
  const form = shallow(<Form>{() => {}}</Form>);
  const formsy = form.find(Formsy.Form);

  formsy.simulate('valid');
  expect(form.state('isValid')).toBeTruthy();
});

it('set isValid to false when formsy is invalid', () => {
  const form = shallow(<Form>{() => {}}</Form>);
  const formsy = form.find(Formsy.Form);

  formsy.simulate('invalid');
  expect(form.state('isValid')).toBeFalsy();
});
