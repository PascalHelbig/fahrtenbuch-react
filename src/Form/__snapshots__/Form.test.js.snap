exports[`test renders correctly 1`] = `
<Formsy
  onChange={[Function]}
  onError={[Function]}
  onInvalid={[Function]}
  onInvalidSubmit={[Function]}
  onSubmit={[Function]}
  onSuccess={[Function]}
  onValid={[Function]}
  onValidSubmit={[Function]}
  preventExternalInvalidation={false}
  validationErrors={null} />
`;
